pkg/apiserver/rest/dualwriter.go

```diff
@@ -73,36 +87,98 @@ func NewDualWriter(legacy LegacyStorage, storage Storage) *DualWriter {
+func (d *DualWriter) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc, forceAllowCreate bool, options *metav1.UpdateOptions) (runtime.Object, bool, error) {
+	// #TODO: figure out which method should be used anytime Get is called
+	// #TODO figure out failure modes, how to guarantee consistency of the updates across both stores
+	// #TODO replace with a rest.CreaterUpdater check?
+	legacy, ok := d.legacy.(rest.Updater)
+	if !ok {
+		return nil, false, fmt.Errorf("legacy storage rest.Updater is missing")
+	}
+
+	// #TODO: Doing it the repetitive way first. Refactor once all the behavior is well defined.
+	switch CurrentMode {
+	case Mode1:
+		_, ok := d.legacy.(rest.Getter)
+		if !ok {
+			return nil, false, fmt.Errorf("legacy storage rest.Getter is missing")
 		}
 
-		accessor, err := meta.Accessor(created)
+		old, err := d.legacy.Get(ctx, name, &metav1.GetOptions{})
 		if err != nil {
-			return created, err
+			return nil, false, err
 		}
-		accessor.SetResourceVersion("")
-		accessor.SetUID("")
 
-		rsp, err := d.Storage.Create(ctx, created, createValidation, options)
+		updated, err := objInfo.UpdatedObject(ctx, old)
 		if err != nil {
-			klog.Error("unable to create object in duplicate storage", "error", err)
+			return nil, false, err
+		}
+		accessor, err := meta.Accessor(updated)
+		if err != nil {
+			return nil, false, err
 		}
-		return rsp, err
-	}
 
-	return d.Storage.Create(ctx, obj, createValidation, options)
-}
+		accessor.SetUID("")
+		accessor.SetResourceVersion("")
+		// #TODO check how much we need to do before legacy.Update for mode 1
+		return legacy.Update(ctx, name, &updateWrapper{
+			upstream: objInfo,
+			updated:  updated,
+		}, createValidation, updateValidation, forceAllowCreate, options)
 
-// Update overrides the default behavior of the Storage and writes to both the LegacyStorage and Storage.
-func (d *DualWriter) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc, forceAllowCreate bool, options *metav1.UpdateOptions) (runtime.Object, bool, error) {
-	if legacy, ok := d.legacy.(rest.Updater); ok {
+	case Mode2:
+		// #TODO figure out how to set opts.IgnoreNotFound = true here
+		// or how to specifically check for an apierrors.NewNotFound error
+		// #TODO for now assume that we are updating entities which had previously
+		// created in entity. GuaranteedUpdate is going to take into account this case:
+		// https://github.com/grafana/grafana/pull/85206
+		// #TODO figure out wht the correct resource version should be and
+		// set it properly in Get and List calls
 		// Get the previous version from k8s storage (the one)
-		old, err := d.Get(ctx, name, &metav1.GetOptions{})
+		old, err := d.Storage.Get(ctx, name, &metav1.GetOptions{})
```

Comment:
```suggestion
		old, err := d.Storage.Get(ctx, name, &metav1.GetOptions{
			IgnoreNotFound: true,
		})
```

Reply:
Thanks for the review. I had tried including `IgnoreNotFound` before, but the field is apparently unknown. I looked into it some more and it looks like it is only available for GetOptions from the [storage package](https://pkg.go.dev/k8s.io/apiserver@v0.29.2/pkg/storage#GetOptions). I'll figure out if/how to update the signature to take this other type.

Reply:
ah yes, we should be passing storage.GetOptions not metav1.GetOptions

Reply:
scratch that, I was forgetting that this is rest, not storage. I guess we just have to catch the error then.